### PR TITLE
DDF-2287: Added integration tests for download interrupt scenarios.

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Fixes the original version of this PR to not depend on sleeps for the test results. This needs serious critique due to the synchronization issues. Background context and explanation to follow. 

#### Who is reviewing it?
@coyotesqrl 
@tbatie 
@gordocanchola 
@rzwiefel 
@brendan-hofmann 
@stustison 
@lessarderic 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@pklinef 
@shaundmorris

#### How should this be tested?
Ensure TestFederation passes. 

#### Any background context you want to provide?
Yes. Will fill out later. Trying to get the code up ASAP. 

* Created the ClientDownloader classes for parallel integration testing.
* Updated async http libraries in test.
* Fix pom dependency to osgi version.
* Cleans up extraneous IntelliJ inspections reported in TestFederation.
* Refactored the entire implementation to use a `CyclicBarrier` instead of pure sleeps.

### CODE REVIEW

**Comment by Lambeaux** [comment]
Bamboo will likely fail. I'm just trying to get feedback on this code as early as possible.

**Comment by Lambeaux** [comment]
Please tell me if I should break this up into multiple PR's to make things easier.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.Collections;
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+import org.glassfish.grizzly.http.server.Response;
+import org.glassfish.grizzly.http.util.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.xebialabs.restito.semantics.Function;
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java]: I believe you meant to use `java.util.function.Function` here.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            byteRange = new ByteRange(requestHeaders.get("range"),
+                    responseMessage.length,
+                    responseMessage);
+            response.setStatus(HttpStatus.PARTIAL_CONTENT_206);
+            response.setHeader("Content-Range", byteRange.contentRangeValue());
+            LOGGER.debug("ChunkedContentResponse: Response range header set to [Content-Range: {}]",
+                    byteRange.contentRangeValue());
+        } else {
+            response.setStatus(HttpStatus.OK_200);
+            byteRange = new ByteRange(0, responseMessage.length - 1, responseMessage);
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java]: The inconsistency between the two ctors bothers me. In one, you provide start/end values necessitating this `-1`; in the other, you provide range values and total size so don't decrement the size. It reads oddly to me.
I'd prefer if both ctors took the total size and your `ByteRange` logic managed the fenceposts.

As I see that these are the only two lines, at least in the current code, that are creating `ByteRange`s, you could manage that simply. In fact, since these are the only creations, you could probably go with something like:
`ByteRange(String rangeHeader, int totalSize, String message)`
and
`ByteRange(int totalSize, String message)`
as your constructors. You seem to always go from 0 to size-1 in the no-header case anyway.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        this.messageDelayMs = messageDelay.toMillis();
+        this.numberOfFailures = numberOfFailures;
+    }
+
+    @Override
+    protected Response send(Response response, ByteRange byteRange) {
+        // send each character, respecting the range header
+        for (int i = byteRange.start; i <= byteRange.end; i++) {
+            try {
+                LOGGER.debug("ChunkedContentResponse: Sending character [{}]", responseMessage[i]);
```
Comment (stustison) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java]: trace

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+                // fail download by ungracefully closing the output buffer to simulate connection lost
+                if (numberOfRetries < numberOfFailures) {
+                    response.getOutputBuffer()
+                            .recycle();
+                    numberOfRetries++;
+                    return response;
+                }
+            } catch (IOException | InterruptedException e) {
+                LOGGER.error("Error", e);
```
Comment (stustison) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java]: Do you have anything better than just "Error"?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                        // If this was triggered, then the next stream.read() should have no
+                        // data available to read. It should block.
+                    }
+                    nextByte = stream.read();
+                }
+            } catch (IOException e) {
+                /* We only warn if the IOException was from a prematurely closed connection
+                    This is an expected use case for the client
+                    Any other cause is an error and a rethrow */
+                e.printStackTrace();
```
Comment (stustison) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java]: no print stack trace

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+
+    public Future<String> begin() {
+        return executorService.submit(() -> {
+            try (InputStream stream = this.inputStream) {
+                int nextByte = stream.read();
+                while (nextByte > -1) {
+                    synchronized (data) {
+                        data.append(nextByte);
```
Comment (stustison) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java]: This would go much faster and use less CPU if you read chunks of bytes instead of one byte at a time.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/DownloadStrategy.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+    static String getAsyncDownloadUrl(String metacardId) {
+        return String.format("%s?source=%s&metacard=%s",
+                AbstractIntegrationTest.RESOURCE_DOWNLOAD_ENDPOINT_ROOT.getUrl(),
+                CSW_STUB_SOURCE_ID,
+                metacardId);
+    }
+
+    static String getDownloadsListUrl() {
+        return String.format("%s/catalog/downloads", AbstractIntegrationTest.SERVICE_ROOT.getUrl());
```
Comment (stustison) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/DownloadStrategy.java]: For something simple like this, I don't feel that `String.format` provides any sort of readability improvement. It also hurts your performance for this operation.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ByteRange.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    ByteRange(String rangeHeaderValue, int totalSizeOfProductInBytes, char[] responseMessage) {
+        this.responseMessage = responseMessage;
+        // extract bytes
+        String startToken = StringUtils.substringBetween(rangeHeaderValue, "=", "-");
+        String endToken = StringUtils.substringAfter(rangeHeaderValue, "-");
+
+        // range offsets can be blank to indicate "until beginning" or "until end" of data.
+        try {
+            if (StringUtils.isBlank(startToken)) {
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ByteRange.java]: It would be clearer if you inverted these tests with `isNotBlank` and assumed the values were set in the primary case.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java (hunk 3 above)
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java]: Itests have logging turned up to DEBUG; this is going to be incredibly noisy (and slow from IO). What value is there in logging at this volume?

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    @Override
+    protected Response send(Response response, ByteRange byteRange) {
+        // send each character, respecting the range header
+        for (int i = byteRange.start; i <= byteRange.end; i++) {
+            try {
+                LOGGER.debug("ChunkedContentResponse: Sending character [{}]", responseMessage[i]);
+                response.getNIOWriter()
+                        .write(responseMessage[i]);
+                response.flush();
+                sleep(messageDelayMs);
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java]: With a sleep between each character, this is going to lead to very long/slow tests. We probably want to establish maximum values for the size of message and length of messageDelay, document them, and throw Exceptions if someone inadvertently writes a test that exceeds them.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+                // fail download by ungracefully closing the output buffer to simulate connection lost
+                if (numberOfRetries < numberOfFailures) {
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java]: Unless `numberOfFailures` is set to `0` on construction, this test will be true every time you run. You'll get one and only one byte and exit.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                    response.getOutputBuffer()
+                            .recycle();
+                    numberOfRetries++;
+                    return response;
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContentFunction.java]: Do you need to `finish()` the response here or is it intentionally unfinished?

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/TriggeredContentFunction.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * @param count the number of bytes to reveal.
+     */
+    public synchronized void revealBytes(int count) {
+        if (count < 1) {
+            throw new IllegalArgumentException("Cannot reveal a negative or zero count of bytes");
+        }
+
+        revealCount = count;
+        senderLock.notify();
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/TriggeredContentFunction.java]: Don't use low-level synchronization.

File: distribution/test/itests/test-itests-ddf/pom.xml [hunk 13]
```diff
@@ -95,14 +95,20 @@
         </dependency>
         <dependency>
             <groupId>org.apache.httpcomponents</groupId>
-            <artifactId>httpcore</artifactId>
-            <version>4.4.5</version>
+            <artifactId>httpcore-osgi</artifactId>
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/pom.xml]: Why this change?

File: distribution/test/itests/test-itests-ddf/pom.xml [hunk 14]
```diff
@@ -95,14 +95,20 @@
...
+            <version>4.4.4</version>
             <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>org.apache.httpcomponents</groupId>
-            <artifactId>httpclient</artifactId>
-            <version>4.5.2</version>
+            <artifactId>httpclient-osgi</artifactId>
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/pom.xml]: Why this change?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 15]
```diff
@@ -2655,9 +2653,104 @@ private CometDClient setupCometDClientWithUser(List<String> channels, String use
...
+            barrier.await(awaitTimeout, TimeUnit.MILLISECONDS);
+            triggeredContentFunction.revealBytes(secondRevealSize);
+            barrier.await(awaitTimeout, TimeUnit.MILLISECONDS);
+            client1.cancelDownload();
+            triggeredContentFunction.revealBytes(lastRevealSize);
+        } catch (TimeoutException e) {
+            fail("Barrier was not triggered within acceptable amount of time");
+        }
+
+        //        WaitCondition.expect("descrip")
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Remove.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 16]
```diff
@@ -2655,9 +2653,104 @@ private CometDClient setupCometDClientWithUser(List<String> channels, String use
...
+        //                .checkEvery(100, TimeUnit.MILLISECONDS)
+        //                .within(5, TimeUnit.SECONDS)
+        //                .until(() -> true);
+
+        String result2 = future2.get();
+
+        logParallelInterruptTestState("[2287]", resourceData, result2);
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: What is this prefix for? What does it signify? How does it clarify output?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 17]
```diff
@@ -2655,9 +2653,104 @@ private CometDClient setupCometDClientWithUser(List<String> channels, String use
...
+        assertThat("Valid result was not complete", result2, is(resourceData));
+    }
+
+    private void logParallelInterruptTestState(String prefix, String... args) {
+        Iterator<String> arg = Arrays.asList(args)
+                .iterator();
+        LOGGER.debug("{} Resource Data: {}", prefix, arg.next());
+        //LOGGER.debug("{} Result 1: {}", prefix, arg.next());
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Remove.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 18]
```diff
@@ -1988,11 +1946,47 @@ public void testCancelDownload() throws Exception {
                 "STOPPED");
     }
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Above this line there are a lot of changes that have nothing to do with your PR (except for the switch to `RestitoContent`).

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    public BlockingInputStreamReader(String metacardId, InputStream inputStream,
+            CyclicBarrier barrier) {
+        super(metacardId, barrier);
+        this.inputStream = inputStream;
+    }
+
+    @Override
+    public void close() throws IOException {
+        inputStream.close();
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java]: It is rare, though not necessarily wrong, for the consumer of an InputStream to be responsible for closing it. Usually it is the responsibility of the originator of the stream to be responsible for its lifecycle.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java (hunk 5 above)
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/BlockingInputStreamReader.java]: Don't print stack trace.

**Comment by coyotesqrl** [comment]
It's hard not to look at this PR and feel that you've over-engineered a bespoke test solution.

My gut says that either you should be able to vastly simplify this or make a more generic, multi-thread test support system. The former would be better at this time.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java (hunk 1 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java]: I don't think restito supports the `java.util.function.Function` - it would be nice if it did.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java (hunk 2 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/AbstractContentFunction.java]: I'll address this. Several classes were extracted into their own files for simplicity. This is not my code but I'll see what I can do to clean it up.

File: distribution/test/itests/test-itests-ddf/pom.xml (hunk 13 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/pom.xml]: It's this or use `wrappedBundle` for my i-test dependencies.

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/TriggeredContentFunction.java (hunk 12 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/TriggeredContentFunction.java]: Is there a compelling alternative? The idea is to guarantee that when `revealBytes` returns, we know the bytes have been sent.

File: distribution/test/itests/test-itests-ddf/pom.xml (hunk 14 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/pom.xml]: It's this or use `wrappedBundle` for my i-test dependencies.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 16 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: This allows me to grep the output using the ticket number.

**Comment by Lambeaux** [comment]
I feel the same way. But I can't think of better options. If we want the same tests without the sleeps, this strikes me as the next-simplest path. Granted it can be cleaned up, but the original issue was over-generification. We could probably take the next step to the latter with some discussion.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 16 above)
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: That doesn't hold long-term value.

**Comment by Lambeaux** [comment]
Per a conversation with @stustison, @lessarderic, and @pklinef we're going to put this on hold until some key refactoring can occur to make things easier to test.
